src/directive.ts

```diff
@@ -0,0 +1,267 @@
+  CHILD: 2,
+  PROPERTY: 3,
+  BOOLEAN_ATTRIBUTE: 4,
+  EVENT: 5,
+  ELEMENT: 6,
+} as const ;
+
+export type PartType = typeof PartType[keyof typeof PartType];
+
+export interface ChildPartInfo {
+  readonly type: typeof PartType.CHILD;
+}
+
+export interface AttributePartInfo {
+  // eslint-disable-next-line @typescript-eslint/type-annotation-spacing
+  readonly type:|typeof PartType.ATTRIBUTE|
+      typeof PartType.PROPERTY|typeof PartType.BOOLEAN_ATTRIBUTE|
+      typeof PartType.EVENT;
+  readonly strings?: ReadonlyArray<string>;
+  readonly name: string;
+  readonly tagName: string;
+}
+
+export type Part = ChildPart|AttributePart|BooleanAttributePart|EventPart;
+
+type Interface<T> = {
+  [P in keyof T]: T[P];
+};
+
+export type ChildPart = Interface<ChildPartImpl>;
+class ChildPartImpl {
+  readonly type = PartType.CHILD;
+  readonly options: legacyLit.RenderOptions|undefined;
+  constructor(readonly legacyPart: legacyLit.NodePart) {
+    this.options = legacyPart.options;
+  }
+
+  get parentNode(): Node {
+    return this.legacyPart.startNode.parentNode!;
+  }
+
+  get startNode(): Node|null {
+    return this.legacyPart.startNode;
+  }
+
+  get endNode(): Node|null {
+    return this.legacyPart.endNode;
+  }
+}
+
+export type AttributePart = Interface<AttributePartImpl>;
+class AttributePartImpl {
+  readonly type: typeof PartType.ATTRIBUTE|typeof PartType.PROPERTY;
+
+  get options(): legacyLit.RenderOptions|undefined {
+    return undefined;
+  }
+  get name(): string {
+    return this.legacyPart.committer.name;
+  }
+
+  get element(): Element {
+    return this.legacyPart.committer.element;
+  }
+
+  /**
+   * If this attribute part represents an interpolation, this contains the
+   * static strings of the interpolation. For single-value, complete bindings,
+   * this is undefined.
+   */
+  get strings() {
+    return this.legacyPart.committer.strings;
+  }
+  get tagName() {
+    return this.element.tagName;
+  }
+
+  constructor(readonly legacyPart: legacyLit.AttributePart|
+              legacyLit.PropertyPart) {
+    if (legacyPart instanceof legacyLit.PropertyPart) {
+      this.type = PartType.PROPERTY;
+    } else {
+      this.type = PartType.ATTRIBUTE;
+    }
+  }
+}
+
+export type BooleanAttributePart = Interface<BooleanAttributePartImpl>;
+class BooleanAttributePartImpl {
+  readonly type = PartType.BOOLEAN_ATTRIBUTE;
+
+  get options(): legacyLit.RenderOptions|undefined {
+    return undefined;
+  }
+  get name(): string {
+    return this.legacyPart.name;
+  }
+
+  get element(): Element {
+    return this.legacyPart.element;
+  }
+
+  /**
+   * If this attribute part represents an interpolation, this contains the
+   * static strings of the interpolation. For single-value, complete bindings,
+   * this is undefined.
+   */
+  get strings() {
+    return this.legacyPart.strings;
+  }
+  get tagName() {
+    return this.element.tagName;
+  }
+
+  constructor(readonly legacyPart: legacyLit.BooleanAttributePart) {
```

Comment:
Don't use constructor parameter properties - declare this as a regular property.

Reply:
Google style guide

Reply:
We don't use Google internal style guide, and we don't use non-standard syntactic features of TypeScript outside of the type system. So no constructor parameter properties, namespaces or enums. Please be consistent with this codebase.